File: 39.组合总和.java

/*
 * @lc app=leetcode.cn id=39 lang=java
 *
 * [39] 组合总和
 */
import java.util.*;

// @lc code=start
class Solution {
    List<List<Integer>> res;
    LinkedList<Integer> path;
    int pathSum = 0;

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        path = new LinkedList<>();
        res = new LinkedList<>();
        backtrack(candidates, 0, target);
        return res;
    }

    private void backtrack(int[] nums, int start, int target) {
        if (pathSum == target) {
            res.add(new LinkedList<>(path));
            return;
        }
        if (pathSum > target)
            return;
        for (int i = start; i < nums.length; i++) {
            pathSum += nums[i];
            path.addLast(nums[i]);
            backtrack(nums, i, target);
            path.removeLast();
            pathSum -= nums[i];
        }
    }
}
// @lc code=end

File: 39.组合总和.cpp
/*
 * @lc app=leetcode.cn id=39 lang=cpp
 *
 * [39] 组合总和
 */
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
// @lc code=start
class Solution
{
public:
    vector<vector<int>> res;
    vector<int> array;
    void FindCombination(int i, vector<int> &candidates, int target)
    {
        if (i == candidates.size())
            return;
        if (target - candidates[i] >= 0)
        {
            array.emplace_back(candidates[i]);
            if (target - candidates[i] == 0)
            {
                res.emplace_back(array);
                array.pop_back();
                // Tree Pruning Here: For a sorted array, if (target - candidates[i] == 0) satisfied,
                // no need to check following elem
                return; // Tree Pruning
            }
            FindCombination(i, candidates, target - candidates[i]);
            array.pop_back();
        }
        FindCombination(i + 1, candidates, target);
    }

    vector<vector<int>> combinationSum(vector<int> &candidates, int target)
    {
        sort(candidates.begin(), candidates.end());
        FindCombination(0, candidates, target);
        return res;
    }
};
// @lc code=end
